bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeChecker.java

```diff
@@ -467,8 +467,30 @@ public static boolean isReferenceEqual(Object lhsValue, Object rhsValue) {
         Type lhsType = getType(lhsValue);
         Type rhsType = getType(rhsValue);
 
-        if (isSimpleBasicType(lhsType) && isSimpleBasicType(rhsType)) {
-            return isEqual(lhsValue, rhsValue);
+        switch(lhsType.getTag()) {
+            case TypeTags.INT_TAG:
+                if (rhsType.getTag() != TypeTags.BYTE_TAG || rhsType.getTag() != TypeTags.INT_TAG) {
+                    return false;
+                }
+                return lhsValue.equals(((Number) rhsValue).longValue());
```

Comment:
Shall we have some tests for both values with same type, in order to cover these lines?